commons/src/main/java/io/aiven/kafka/tieredstorage/commons/SegmentManifestProvider.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.commons;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.commons.manifest.SegmentManifest;
+import io.aiven.kafka.tieredstorage.commons.storage.FileFetcher;
+import io.aiven.kafka.tieredstorage.commons.storage.StorageBackEndException;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+class SegmentManifestProvider {
+    private final ObjectKey objectKey;
+    private final Cache<String, SegmentManifest> cache;
+    private final FileFetcher fileFetcher;
+    private final ObjectMapper mapper;
+
+    /**
+     * @param maxCacheSize the max cache size (in items) or {@code null} if the cache is unbounded.
+     * @param cacheRetention the retention time of items in the cache or {@code null} if infinite retention.
+     */
+    SegmentManifestProvider(final ObjectKey objectKey,
+                            final Long maxCacheSize,
+                            final Duration cacheRetention,
+                            final FileFetcher fileFetcher,
+                            final ObjectMapper mapper) {
+        this.objectKey = objectKey;
+        final var cacheBuilder = Caffeine.newBuilder();
+        if (maxCacheSize != null) {
```

Comment:
Am I missing something or instead we could check for -1 here instead of checking twice, first for -1 and then for null?

Reply:
We're forced to use `-1`: there's no way (that I'm aware of) to explicitly pass `null` from a config file. Kafka uses `-1` as a substitute and I follow suit. However, then I prefer to convert this `-1` to something more proper that conveys special meaning, which is `null` in this case.

Reply:
In the cache implementation -1 is a constant called `UNSET_INT`, maybe we should use that approach to avoid having 2 ifs instead of 1?

Reply:
You cannot pass `-1`, it'll fail with `maximum size must not be negative`

Reply:
Looks like there is a total misunderstanding here. When I was talking about 2 if statements I meant this https://github.com/aiven/tiered-storage-for-apache-kafka/blob/644fadb0dea021240210eaed5a8163bcdb7fad4c/commons/src/main/java/io/aiven/kafka/tieredstorage/commons/UniversalRemoteStorageManagerConfig.java#L194
We are checking for -1 there and returning `null` and then we are checking for `null` here. So my point was that we can pass the value without check from config to here and here have a constant like `UNSET_INT` with the value `-1` to check if the value from config is equal to `UNSET_INT`
I hope now it's clear enough and yes, I'm aware of the fact that -1 can not be passed as an argument. Also the same is probably applicable to `cacheRetention`.

Reply:
Yeah, I got you point.
What I mean is this: At the moment, the special meaning for `-1` exists only in the config definition file. OTOH `null` is universally used for "no value". So I chose here to deal with `-1` no more than necessary and pass `null` around instead.
If we don't do this, at all the places where `maxCacheSize` is used, we have to remember (carry it in some documentation or our heads) that there's some special meaning for `-1`, which we need to check. 3 lines of a trivial `if` is a small price to pay for not having to do this.
Admittedly, this case is simple and the difference is not clear cut. But it's just basic hygiene.

Reply:
Why not to use Optional if that's the intent? That will clearly say that we need to check for value presence.

Reply:
Yes, let me switch to `Optional`

Reply:
Done

Reply:
Or well, it's clear why not to. But I still do not see a big difference in remembering to check for -1 or for null since -1 has a common lack of value interpretation for primitive arguments like sizes or durations. As for spreading the checks across the code with `maxCacheSize` usage it will anyways take place but with null unless we create a generic wrapper for Caffeine and do it in a single place and I will try to do it some point. Lets not spend more time on this one and as I said I will try to address this later in another PR.